app/scripts/modules/core/src/pipeline/config/stages/producesArtifacts/producesArtifacts.component.ts

```diff
@@ -0,0 +1,66 @@
+import { IComponentOptions, IController, module } from 'angular';
+
+import { UUIDGenerator } from 'core/utils/uuid.service';
+import { IStage, IExpectedArtifact } from 'core';
+
+class ProducesArtifactsCtrl implements IController {
+  public stage: IStage;
+
+  public removeExpectedArtifact(stage: IStage, expectedArtifact: IExpectedArtifact) {
+    stage.expectedArtifacts = stage.expectedArtifacts
+      .filter((a: IExpectedArtifact) => a.id !== expectedArtifact.id);
+  }
+
+  private defaultArtifact() {
+    return { kind: 'custom' };
+  }
+
+  public addExpectedArtifact() {
+    const newArtifact = {
+      matchArtifact: this.defaultArtifact(),
+      usePriorExecution: false,
+      useDefaultArtifact: false,
+      defaultArtifact: this.defaultArtifact(),
+      id: UUIDGenerator.generateUuid()
+    };
+
+    if (!this.stage.expectedArtifacts) {
+      this.stage.expectedArtifacts = [];
+    }
+
+    this.stage.expectedArtifacts.push(newArtifact);
+  }
+}
+
+class ProducesArtifactsComponent implements IComponentOptions {
+  public bindings: any = { stage: '=' };
+  public controllerAs = 'ctrl';
+  public controller = ProducesArtifactsCtrl;
+  public template = `
+    <div class="container-fluid form-horizontal">
+      <expected-artifact
+        ng-repeat="expectedArtifact in ctrl.stage.expectedArtifacts"
+        remove-expected-artifact="ctrl.removeExpectedArtifact"
+        context="ctrl.stage"
+        expected-artifact="expectedArtifact"
+        application="application">
+      </expected-artifact>
+      <div class="row" ng-if="!ctrl.stage.expectedArtifacts.length">
```

Comment:
can this explode if `ctrl.stage.expectedArtifacts` is not an array?

we dont check for this on lines 10-11 in `removeExpectedArtifact()` but we do check for it on lines 27-29 in `addExpectedArtifact()`.  should that same check be appearing here in this `ng-if` and in `removeExpectedArtifact()` as well (or maybe only in the constructor of `ProducesArtifactsCtrl` to avoid repetition?)

Reply:
Good catch